core/src/test/scala/unit/kafka/admin/AdminRackAwareTest.scala

```diff
@@ -0,0 +1,228 @@
+  }
+
+  @Test
+  def testLargeNumberPartitionsAssignment() {
+    val numPartitions = 96
+    val replicationFactor = 3
+    val brokerRackMapping = Map(0 -> "rack1", 1 -> "rack2", 2 -> "rack2", 3 -> "rack3", 4 -> "rack3", 5 -> "rack1",
+      6 -> "rack1", 7 -> "rack2", 8 -> "rack2", 9 -> "rack3", 10 -> "rack1", 11 -> "rack3")
+    val assignment = AdminUtils.assignReplicasToBrokers(toBrokerMetadata(brokerRackMapping), numPartitions,
+      replicationFactor)
+    checkReplicaDistribution(assignment, brokerRackMapping, brokerRackMapping.size, numPartitions,
+      replicationFactor)
+  }
+
+  @Test
+  def testMoreReplicasThanRacks() {
+    val numPartitions = 6
+    val replicationFactor = 5
+    val brokerRackMapping = Map(0 -> "rack1", 1 -> "rack2", 2 -> "rack2", 3 -> "rack3", 4 -> "rack3", 5 -> "rack2")
+    val assignment = AdminUtils.assignReplicasToBrokers(toBrokerMetadata(brokerRackMapping), numPartitions, replicationFactor)
+    assertEquals(List.fill(assignment.size)(replicationFactor), assignment.values.map(_.size))
+    val distribution = getReplicaDistribution(assignment, brokerRackMapping)
+    for (partition <- 0 until numPartitions)
+      assertEquals(3, distribution.partitionRacks(partition).toSet.size)
+  }
+
+  @Test
+  def testLessReplicasThanRacks() {
+    val numPartitions = 6
+    val replicationFactor = 2
+    val brokerRackMapping = Map(0 -> "rack1", 1 -> "rack2", 2 -> "rack2", 3 -> "rack3", 4 -> "rack3", 5 -> "rack2")
+    val assignment = AdminUtils.assignReplicasToBrokers(toBrokerMetadata(brokerRackMapping), numPartitions,
+      replicationFactor)
+    assertEquals(List.fill(assignment.size)(replicationFactor), assignment.values.map(_.size))
+    val distribution = getReplicaDistribution(assignment, brokerRackMapping)
+    for (partition <- 0 to 5)
+      assertEquals(2, distribution.partitionRacks(partition).toSet.size)
+  }
+
+  @Test
+  def testSingleRack() {
+    val numPartitions = 6
+    val replicationFactor = 3
+    val brokerRackMapping = Map(0 -> "rack1", 1 -> "rack1", 2 -> "rack1", 3 -> "rack1", 4 -> "rack1", 5 -> "rack1")
+    val assignment = AdminUtils.assignReplicasToBrokers(toBrokerMetadata(brokerRackMapping), numPartitions, replicationFactor)
+    assertEquals(List.fill(assignment.size)(replicationFactor), assignment.values.map(_.size))
+    val distribution = getReplicaDistribution(assignment, brokerRackMapping)
+    for (partition <- 0 until numPartitions)
+      assertEquals(1, distribution.partitionRacks(partition).toSet.size)
+    for (broker <- brokerRackMapping.keys)
+      assertEquals(1, distribution.brokerLeaderCount(broker))
+  }
+
+  @Test
+  def testSkipBrokerWithReplicaAlreadyAssigned() {
+    val rackInfo = Map(0 -> "a", 1 -> "b", 2 -> "c", 3 -> "a", 4 -> "a")
+    val brokerList = 0 to 4
+    val numPartitions = 6
+    val replicationFactor = 4
+    val brokerMetadatas = toBrokerMetadata(rackInfo)
+    assertEquals(brokerList, brokerMetadatas.map(_.id))
```

Comment:
Nitpick: we don't really need this `assertEquals` or the `brokerList` val since that is checking that `toBrokerMetadata` works correctly, which is not the purpose of this test.